### DISCUSSION TITLE
generateStaticParams ~~> TypeError: result is not iterable

### DISCUSSION THREAD

**Discussion by GetWrecked99** [vercel/next.js#54097] [Help] [answered]
### Summary

hey there , i'm using next js v13.4.10 & app direcotry. 
im getting this error while using generateStaticParams function for each api endpoints : 
- error Failed to generate static paths for /study/[slug]/study-guide: TypeError: result is not iterable

this is my implementation : 

![image](https://github.com/vercel/next.js/assets/124883615/7facccb2-5696-43d6-ae5f-01edf01a138f)

**Comment by icyJoseph** [comment, marked as answer]
Await the `.json` call:

```js
const res = await getCountries.json();
```

And also map returning objects by doing:

```js
return res.map((item) => ({ slug: item.id}))
```

In short, you have to return from the `generateStaticParams` function.

**Reply by GetWrecked99** [reply]
i done that but it does not make an effect, still getting the same error

**Reply by icyJoseph** [reply]
Right and then you have some commented code that I ignored.

You have to return this:

```js
const res = await getCountries.json();

return res.map((item) => ({ slug: item.id}))
```

You are not returning anything as you map by doing `(item) => {slug: item.id}`, to return an object in shorthand syntax you gotta do `(item) => ({ /* object contents */ })`

**Reply by icyJoseph** [reply]
And now I see that the API you are using does not have any such `id` property on the items it returns...

**Reply by icyJoseph** [reply]
And if you don't want to map, at all, then just do:

```js
const res = await fetch("https://restcountries.com/v3.1/all");
const result = await res.json();
return result;
```

Though that has a bunch of unwanted properties, you probably don't want to use. You should map and pick only what you need.

**Reply by GetWrecked99** [reply]
i just switched my api and In addition to the previous error, i got this message result variable on the terminal console : 
{ detail: 'Not authenticated' }
so , i tried to send the token like this but getting new error : 

![image](https://github.com/vercel/next.js/assets/124883615/fcf98e87-6385-467d-b19b-301da5e536aa)

**Reply by icyJoseph** [reply]
Why don't you return from your `generateStaticParams` function? You have to return something at the end, like:

```js
return result;
```

That's the root issue you are having. Again your function must return something.

Regarding the second error, I am not sure, is the screenshot from a Page component? It feels like it is getting imported into a client component somehow?

**Reply by GetWrecked99** [reply]
well, i am using this implementation in a server component (app dir). 
this kind of cookie implementation is working correctly inside of my server components body, but i have this error in the generateStaticParams function (something like global scope issues or etc...) 
firstly i believe that we have to pass the token somehow then see the other problems 

![image](https://github.com/vercel/next.js/assets/124883615/bbb157c5-60b5-4e29-8005-fbe1a1ca5ba0)

**Reply by icyJoseph** [reply]
Well the token seems to be correct. Assuming the token exists. You can `console.log(token)` and that should be an object with name and value keys.

## Edit

The issue here is that cookies is a dynamic value, and while rendering the paths, we need to be able to calculate these at build time, and there's no cookies there, obviously.

~Then it should work, but... The problem with `requestAsyncStorage` seems to be Windows related. I found this ongoing issue: https://github.com/vercel/next.js/issues/52176, where other people run into the same problem.~

**Reply by icyJoseph** [reply]
~Another issue https://github.com/vercel/commerce/issues/1164 - looks like windows is to blame for `requestAsyncStorage` problems.~

**Reply by icyJoseph** [reply]
Aha wait, using cookies within `getStaticParams` is not gonna fly... that's the issue... 🤔 🤦 

Your token should be an env var or something like that.

Does `penseum` provide an API key or something you could use to authenticate?

**Reply by GetWrecked99** [reply]
well, the only thing i have access to authenticate a user is a simple login api endpoint. when the user just logged in into the app, it gives me the access token. nothing more, nothing less. 
i do not have a static token key for making a var/env for it

**Reply by icyJoseph** [reply]
Ok, so we are now solving a different issue. As far as your original question goes, I believe it has been answered by:

- https://github.com/vercel/next.js/discussions/54097#discussioncomment-6738001

I am not sure how you gain access to the Penseum API, but assuming they put a token as a cookie in the browser, we could do this modification (only conceptually - do not copy this code):

```tsx
// pseudo-code
const StudyGuidePage  = ({params}) => {
  const {slug} = params;
  const token = cookies().get("token")?.value; // can be null
  // get the courses as you do
  const {courses} = await res.json();
  // no course matches the slug
  if (!courses.some(course => course.course_id === slug)) {
    notFound(); // imported from 'next/navigation'
  }
  const data = /*  get data */
  return /* component */
}
```

This is incredibly bad of course, because, for every attempt at seeing a page slug, we get all courses, map over than and make a conclusion.

So the question is, could your `getCourseContentApi` function, return `null` when the `slug` has no content, for example when it 404's? And then:

```jsx
  const data = /*  get data */
  if (data === null) {
   notFound(); // imported from 'next/navigation'
  }
  return /* component */
```

The problem being that you can't pre-build the paths, because they depend on a runtime dynamic value, which is the token cookie. `generateStaticParams` is useful only when you know everything upfront during build time. In this case, you need the token to be able to do anything at all. So you need to render as the request comes in. Similar to what `GetServerSideProps` used to do.

This second approach is better because each attempt results in one request, and that's it.

**Reply by GetWrecked99** [reply]
yup , i got that. thank you for replying <3 have a good time
